### ISSUE TITLE
last subtitle text appears forever

### ISSUE THREAD

**Issue by amuth** [gpac/gpac#362]
Hello

If I add a subtitle, the last subtitle text appears forever at the end of the film. Have tested it with KMPlayer, VLC and MPC. Always the same behavior.

Command-line: mp4box.exe -add "sw2.264" -add "sw2.aac" -add "sw2.srt" -new "sw2.mp4"

Testet with the last version: gpac-0.5.2-DEV-rev1110

**Comment by combucha** [comment]
Same here, nothing new, old builds behave the same way.
Anyway, I tried to "kill" last textline by inserting space with timecodes and mp4box skipped this line.
It ignores importing this space (or spaces) also if placed in any part of .srt file. Exported back -raw .tx3g files have this lines missing.
Importing .srt with mp4box needs revision.

**Comment by rbouqueau** [comment]
Could it be related to #95 ?

**Comment by combucha** [comment]
Timecodes exported back to .srt are ok.
I'm not sure if this displayed forever last subtitle is really mp4box issue.
VLC & MPC-HC indeed show last textline till end of movie, but if I touch seek fwd it vanish in VLC, but stay in MPC-HC. Other player MPC-BE however plays such files correctly.

Imo, mp4box should inject .srt into mp4 as it was created, without parsing "spacelines" mentioned earlier. I know they are invisible...  ;)

**Comment by amuth** [comment]
ITunes will also correctly play subtitles. I think the problem is not with mp4box.

Thank you for your help

**Comment by dustsigns** [comment]
I (original author of issue #95) second @combucha's suggestion to keep the empty lines ("space lines") to not break playback in MPC-HC and VLC (partially). It would be nicer to have the old behavior (of MP4Box version 0.46 and prior, see issue #95) back.
Note that I did not reopen issue #95 although it reappeared for my files - this issue and issue #95 are identical. For reproducing the issue you need a player which is sensitive to the SRT formatting as described by @combucha above.

**Comment by rbouqueau** [comment]
Thanks all for your contributions. #95 was closed because we couldn't reproduce. Would any of you be able to provide a sample content and a list of players that we should test on?

**Comment by dustsigns** [comment]
Step 1: Take a video, e.g., http://temp.dustsigns.de/test.264
Step 2: Take a subtitle file where the last subtitle ends well before the end of the video, e.g., http://temp.dustsigns.de/test.srt
Step 3: Mux them with a current version of mp4box (from http://download.tsi.telecom-paristech.fr/gpac/latest_builds/win64/gpac-0.6.0-DEV-rev9-g5d67fa4-master-x64.exe): `"C:\Program Files\GPAC\mp4box.exe" -fps 25 -add test.264 -add test.srt:lang=eng:name=Test -new broken.mp4`
Step 4: Play the MP4 file with a current version of MPC HC (from https://binaries.mpc-hc.org/MPC%20HomeCinema%20-%20x64/MPC-HC_v1.7.10_x64/MPC-HC.1.7.10.x64.exe)
Step 5: Watch as the last subtitle does not disappear, but stays until the end of the playback

If you repeat the steps with v0.46 of mp4box in step 3 instead, everything works as expected, i.e., the last subtitle disappears when it should.

"Appendix":
In case the files above are taken down at some point, here is how they can be reproduced:

test.264:
Step 1: Write

> BlankClip(length=1000, width=640, height=480, fps=25)

to test.avs
Step 2: Encode test.avs with an x264 version with AviSynth support to get test.264 (I was lazy here and used StaxRip which makes use of an intermediate tool like this: `"C:\StaxRip\Applications\avs4x26x\avs4x26x.exe" --x26x-binary "C:\StaxRip\Applications\x264\x264 64-Bit 8-Bit.exe" --crf 22 --sar 255:254 --output test.264 test.avs`)

test.srt:
> 1
> 00:00:01,000 --> 00:00:04,000
> A subtitle
> 
> 2
> 00:00:05,000 --> 00:00:09,000
> Another subtitle

Screenshots to show what happens using this example: http://temp.dustsigns.de/broken.png (current mp4box version) vs. http://temp.dustsigns.de/test.png (version 0.46 of mp4box).

**Comment by cconcolato** [comment]
v0.6.0 creates a track with 4 samples and with a duration of 9 seconds. 
```
Sample 1	DTS 0	CTS 0	2	1	14691	0	0	0	0	0	0	0
Sample 2	DTS 1000	CTS 1000	12	1	14693	0	0	0	0	0	0	0
Sample 3	DTS 4000	CTS 4000	2	1	14978	0	0	0	0	0	0	0
Sample 4	DTS 5000	CTS 5000	18	1	15253	0	0	0	0	0	0	0
```
v0.4.6 creates a track with 5 samples and the track duration is 13 seconds. 
```
Sample 1	DTS 0	CTS 0	2	1	14619	0	0	0	0	0	0	0
Sample 2	DTS 1000	CTS 1000	12	1	14621	0	0	0	0	0	0	0
Sample 3	DTS 4000	CTS 4000	2	1	14906	0	0	0	0	0	0	0
Sample 4	DTS 5000	CTS 5000	18	1	15181	0	0	0	0	0	0	0
Sample 5	DTS 9000	CTS 9000	2	1	15472	0	0	0	0	0	0	0
```

It's unclear to me why v0.4.6 was adding this extra sample nor why that sample was made 4 secs long. This could be problematic if you concatenate MP4 files, as you wouldn't expect this 4 seconds gap between the two files. 

The 0.6.0 version is correct and players should be fixed to support it. I'm reluctant to change the current behavior, at least not as the default behavior. Maybe we could use the `:dur` parameter and if the dur is actually longer than the content duration, we could force an "empty" subtitle.

**Comment by dustsigns** [comment]
As @combucha wrote above, this affects VLC to a certain degree as well and may (also?) be due to the (last?) empty line being removed. Would it be a possibility to keep this line, if it is causing two very frequently used media players such problems? If the empty lines are not the actual reason, I still find it quite strange that muxing an SRT causes this unexpected behavior in two different (and very frequently used) media players.

**Comment by rbouqueau** [comment]
@dustsigns For our information are there parallel discussions with VLC and MPC-HC maintainers?

Possible commits related to this issue:
 - 9123bb4c4fc587af0f4a72c03b73250466afae9d
 - ee33e7f374f832ac7d85b75fca5cd37fe495377a

**Comment by rbouqueau** [comment]
And 6749be9f42305ede00c373dce5011dae405ca139

**Comment by cconcolato** [comment]
@dustsigns I understand your concern but @combucha indicates that seeking wipes out this last line. So this is definitely a bug in the players. You should file a bug with those players.

There is no "last line being removed". The content in the MP4 is the same as the content in the SRT as the export shows. If we add anything to the MP4, the export in the SRT will differ from the initial one. This would be wrong unless the user says so when it packages the SRT in the MP4. We cannot "keep" this line since it is not there. As I suggested, I'm willing to "add" it through a parameter in the command line but not as the default.

**Comment by dustsigns** [comment]
@rbouqueau: No, there are no discussions at this moment that I am aware of.
@cconcolato: Not only seeking wipes out the last line. Simply playing the example file I described above leads to the error in question in MPC HC.

**Comment by rbouqueau** [comment]
2 points:
- Thanks for the sample, I could reproduce. I am looking for another muxer that would work with these players without inserting the last empty sample. I haven't found any. Any idea?
- It also seems to me that MP4Box is right so it might be useful to report to the players. Maybe they also have ideas on the subject.

Thanks

**Comment by dustsigns** [comment]
It seems that avconv does not exhibit this issue. I tried the following on my Ubuntu machine:

`avconv -y -i test.264 -i test.srt -codec copy test.mp4`

Both, MPCHC and VLC play the MP4 file just fine. However, I am not entirely sure whether avconv actually leaves the SRT untouched, i.e., whether it muxes it (instead of transcoding to some other subtitle format). It seems to be muxing, but I cannot currently verify this (at work).

**Comment by rbouqueau** [comment]
Thank you. I cannot reproduce with both FFmpeg and libav with the same command-line. Could you please upload you resulting file?

PS: we have a dropbox if useful: https://www.mediafire.com/filedrop/filedrop_hosted.php?drop=eec9e058a9486fe4e99c33021481d9e1826ca9dbc242a6cfaab0fe95da5e5d95

**Comment by dustsigns** [comment]
I uploaded the file to http://temp.dustsigns.de/test.mp4

Details which may be necessary to reproduce the file:
`uname -rv` yields 

> 3.13.0-74-generic #118-Ubuntu SMP Thu Dec 17 22:52:10 UTC 2015

`avconv` yields

> avconv version 9.18-6:9.18-0ubuntu0.14.04.1, Copyright (c) 2000-2014 the Libav developers
  built on Mar 16 2015 13:19:10 with gcc 4.8 (Ubuntu 4.8.2-19ubuntu1)
[..]

**Comment by cconcolato** [comment]
That file only has one video track, no subtitle track. I suspect the subtitles are embedded in the video track (CEA 608, not burnt in) because mp4client plays a dark video while VLC sees a subtitle stream.

**Comment by dustsigns** [comment]
Although I think I finally found the correct muxing command, unfortunately, it seems that the avconv version that I am using does not support SRT in MP4:
`avconv -y -i test.264 -i test.srt -map 0 -map 1 -codec copy test.mp4` outputs
> [..]
Input #1, srt, from 'test.srt':
  Duration: N/A, bitrate: N/A
    Stream #1.0: Subtitle: srt
[mp4 @ 0x1e27ee0] track 1: could not find tag, codec not currently supported in container
[..]

Is anyone aware of any other muxers apart from avconv (and mp4box, of course) that we could try?

**Comment by cconcolato** [comment]
Correcting my previous message. VLC does not play subtitles from the file you gave (it was finding the srt file in the same folder ....).

**Comment by rbouqueau** [comment]
> 2 points:
> 
> - Thanks for the sample, I could reproduce. I am looking for another muxer that would work with these players without inserting the last empty sample. I haven't found any. Any idea?

We found none at this point.

When I played with FFmpeg I could burn the subtitles, but not mux them in the MP4.

> It also seems to me that MP4Box is right so it might be useful to report to the players. Maybe they also have ideas on the subject.

Could someone report to VLC (https://trac.videolan.org/vlc/) and MPC-HC (https://trac.mpc-hc.org/wiki/How_to_Report_Issues) please? I think that would help a lot. Thank you.

**Comment by dustsigns** [comment]
I reported the issue in the MPC-HC tracker: https://trac.mpc-hc.org/ticket/5781#ticket. I also uploaded the MP4 file (broken.mp4) described in my post above there so that the issue is easier to reproduce for them.

**Comment by Underground78** [comment]
The problem isn't on MPC-HC side as far as I can tell. Our subtitle renderer doesn't get the right stop time for the last subtitle entry. I would guess the problem is within ffmpeg's demuxer if you're sure that the file is properly muxed.

**Comment by rbouqueau** [comment]
Thanks. After a discussion with @jeanlf I have created a pull request to go back to the previous behaviour.

Rationale: we already modify the existing content by inserting empty samples.

We need to check that GPAC behaviour is consistent across subs formats.

**Comment by rbouqueau** [comment]
Merged: 6671342412ba60fc615c5b760d0a15447828e2ee

Please re-open issue if the bug is still present.
